### PR TITLE
feat!: provider proposal for changing reward denoms

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
## Description

Note I don't think there's any migrations required for a PR that adds a new gov proposal type and removes an existing tx. 

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if state-machine breaking change (i.e., requires coordinated upgrade)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [x] Added a changelog entry to `CHANGELOG.md`
* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [x] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [x] Confirmed all CI checks have passed
* [x] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release for both the consumer and provider

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

**Review by mpoke — APPROVED** [review]
Great work @smarshall-spitzbart

**Review by p-offtermatt — COMMENTED** [review]
https://github.com/cosmos/interchain-security/blob/3e823c3c2bf9fc85a0f0ee8f577e6aad4c3f1627/x/ccv/provider/types/msg.go#L13 The MsgRegisterConsumerRewardDenom type seems unused

File: x/ccv/provider/types/proposal.go [hunk 1]
```diff
@@ -231,3 +234,44 @@ func (sp *EquivocationProposal) ValidateBasic() error {
...
+// ProposalRoute returns the routing key of a change reward denoms proposal.
+func (crdp *ChangeRewardDenomsProposal) ProposalRoute() string { return RouterKey }
+
+// ProposalType returns the type of a change reward denoms proposal.
+func (crdp *ChangeRewardDenomsProposal) ProposalType() string {
+	return ProposalTypeChangeRewardDenoms
+}
+
+// ValidateBasic runs basic stateless validity checks on a ChangeRewardDenomsProposal.
+func (crdp *ChangeRewardDenomsProposal) ValidateBasic() error {
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/types/proposal.go:259]: https://github.com/cosmos/interchain-security/commit/c998866b9c2142124a0574937499e1c9b411ab78

File: x/ccv/provider/keeper/proposal.go [hunk 2]
```diff
@@ -616,3 +616,31 @@ func (k Keeper) HandleEquivocationProposal(ctx sdk.Context, p *types.Equivocatio
 	}
 	return nil
 }
+
+func (k Keeper) HandleConsumerRewardDenomProposal(ctx sdk.Context, p *types.ChangeRewardDenomsProposal) error {
+	if err := p.ValidateBasic(); err != nil {
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/proposal.go]: https://github.com/cosmos/interchain-security/commit/4d03abddc0a3b2d929033e6febd027615109f5e1

**Comment by smarshall-spitzbart** [comment]
> https://github.com/cosmos/interchain-security/blob/3e823c3c2bf9fc85a0f0ee8f577e6aad4c3f1627/x/ccv/provider/types/msg.go#L13
> 
> The MsgRegisterConsumerRewardDenom type seems unused

Thanks, removed in https://github.com/cosmos/interchain-security/commit/129f95972ea905bd3f873d7d23989185cb8acc14

File: tests/e2e/actions.go [hunk 3]
```diff
@@ -1751,35 +1751,64 @@ func (tr TestRun) registerRepresentative(
 	wg.Wait()
 }
 
-type registerConsumerRewardDenomAction struct {
-	chain chainID
-	from  validatorID
-	denom string
+type submitChangeRewardDenomsProposalAction struct {
+	chain   chainID
```
Comment (p-offtermatt) [inline tests/e2e/actions.go:1755]: Could you make the name more descriptive? This is the consumer chain, I assume?

Comment (smarshall-spitzbart) [reply tests/e2e/actions.go]: I just removed it, only used in a dummy title anyways :)

https://github.com/cosmos/interchain-security/commit/c4e4a6945ec7c58bf87d8bcce8a18848de00748a

It was the chain in which the prop is executed, which is provider

File: tests/e2e/actions.go [hunk 4]
```diff
@@ -1751,35 +1751,64 @@ func (tr TestRun) registerRepresentative(
...
-		fmt.Println("redelegate cmd:", string(bz))
-	}
-
 	if err != nil {
 		log.Fatal(err, "\n", string(bz))
 	}
 
-	tr.waitBlocks(action.chain, 2, 10*time.Second)
+	// wait for inclusion in a block -> '--broadcast-mode block' is deprecated
+	tr.waitBlocks(chainID("provi"), 2, 30*time.Second)
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply tests/e2e/actions.go:1810]: Either way could work, but action.chain is removed now

File: tests/e2e/actions.go (hunk 3 above)
Comment (p-offtermatt) [reply tests/e2e/actions.go]: Thanks, makes sense!

File: x/ccv/provider/client/proposal_handler.go [hunk 5]
```diff
@@ -337,18 +410,28 @@ func ParseEquivocationProposalJSON(proposalFile string) (EquivocationProposalJSO
 	return proposal, nil
 }
 
-func ParseConsumerRemovalProposalJSON(proposalFile string) (ConsumerRemovalProposalJSON, error) {
-	proposal := ConsumerRemovalProposalJSON{}
+type ChangeRewardDenomsProposalJSON struct {
+	Summary string `json:"summary"`
+	types.ChangeRewardDenomsProposal
```
Comment (p-offtermatt) [inline x/ccv/provider/client/proposal_handler.go:415]: This seems like a much cleaner pattern than the other proposal types, nice!

**Review by p-offtermatt — APPROVED** [review]
Thanks for the work, looks great! A couple of nitpicks, but I'm approving since nothing is truly problematic

File: proto/interchain_security/ccv/provider/v1/tx.proto [hunk 6]
```diff
@@ -30,18 +28,3 @@ message MsgAssignConsumerKey {
 }
 
 message MsgAssignConsumerKeyResponse {}
-
-// MsgRegisterConsumerRewardDenom allows an account to register
-// a consumer reward denom, i.e., add it to the list of denoms
-// accepted by the provider as rewards.
-message MsgRegisterConsumerRewardDenom {
```
Comment (MSalopek) [inline proto/interchain_security/ccv/provider/v1/tx.proto:37]: Note: Remember to tag the protos on https://buf.build/cosmos/interchain-security with the specific version this ends up in.

File: x/ccv/provider/keeper/proposal.go [hunk 7]
```diff
@@ -616,3 +616,28 @@ func (k Keeper) HandleEquivocationProposal(ctx sdk.Context, p *types.Equivocatio
 	}
 	return nil
 }
+
+func (k Keeper) HandleConsumerRewardDenomProposal(ctx sdk.Context, p *types.ChangeRewardDenomsProposal) error {
```
Comment (MSalopek) [inline x/ccv/provider/keeper/proposal.go:620]: Run `make lint`

File: x/ccv/provider/types/proposal_test.go [hunk 8]
```diff
@@ -362,3 +362,38 @@ func TestEquivocationProposalValidateBasic(t *testing.T) {
...
+		name: "invalid change reward denoms proposal, none to add or remove",
+		proposal: types.NewChangeRewardDenomsProposal(
+			"title", "description", []string{}, []string{}),
+		expectError: true,
+	}, {
+		name: "invalid change reward denoms proposal, same denom in both sets",
+		proposal: types.NewChangeRewardDenomsProposal(
+			"title", "description", []string{"denom1"}, []string{"denom1"}),
+		expectError: true,
+	}, {
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/types/proposal_test.go]: Good call, https://github.com/cosmos/interchain-security/commit/93b68a95d75829dd0392d9343638d718747ade7e

File: x/ccv/provider/keeper/proposal.go [hunk 9]
```diff
@@ -616,3 +616,28 @@ func (k Keeper) HandleEquivocationProposal(ctx sdk.Context, p *types.Equivocatio
...
+			sdk.NewAttribute(ccv.AttributeConsumerRewardDenom, denomToAdd),
+		))
+	}
+	for _, denomToRemove := range p.DenomsToRemove {
+		// Log error and move on if one of the denoms is not registered
+		if !k.ConsumerRewardDenomExists(ctx, denomToRemove) {
+			ctx.Logger().Error("denom %s not registered", denomToRemove)
+			continue
+		}
+		k.DeleteConsumerRewardDenom(ctx, denomToRemove)
```
[the remark this answers is not in the document]
Comment (smarshall-spitzbart) [reply x/ccv/provider/keeper/proposal.go:639]: Nice catch, https://github.com/cosmos/interchain-security/commit/a2ed4df7e9d285ff9747a939290014eb486acaf1

**Pull request merged**